### PR TITLE
JCLOUDS-1587 S3 Blobstore: Support Bucket Configuration Options - Versioning, Lifecycle, Encryption

### PR DESCRIPTION
**Pull request by jiffysrc**
https://issues.apache.org/jira/browse/JCLOUDS-1587

Support following features in jclouds s3 blobstore

Configure the bucket for versioning and get the versioning state of a bucket
Confiure the bucket for encryption and get the encryption state of a bucket
Confiure the bucket for lifecycle rules and get the lifecycle rules of a bucket
Fetch an object using specific versionId
List different versions of an Object

### CODE REVIEW

File: apis/s3/src/main/java/org/jclouds/s3/S3Client.java [hunk 1]
```diff
@@ -260,7 +264,7 @@ DeleteResult deleteObjects(@Bucket @EndpointParam(parser = AssignCorrectHostname
     * @return ETag of the content uploaded
     * @throws org.jclouds.http.HttpResponseException
     *            if the conditions requested set are not satisfied by the object on the server.
-    * @see org.jclouds.s3.domain.CannedAccessPolicy#PRIVATE
+    * @see CannedAccessPolicy#PRIVATE
```
Comment (gaul) [inline apis/s3/src/main/java/org/jclouds/s3/S3Client.java:267]: There are a lot of noise changes in this PR -- can you revert them?

File: apis/s3/src/main/java/org/jclouds/s3/blobstore/S3BlobStore.java [hunk 2]
```diff
@@ -174,7 +189,12 @@ public void setContainerAccess(String container, ContainerAccess access) {
    @Override
    public PageSet<? extends StorageMetadata> list(String container, ListContainerOptions options) {
       ListBucketOptions httpOptions = container2BucketListOptions.apply(options);
-      PageSet<? extends StorageMetadata> list = bucket2ResourceList.apply(sync.listBucket(container, httpOptions));
+      PageSet<? extends StorageMetadata> list = null;
+      if (httpOptions.getVersions()){
```
Comment (gaul) [inline apis/s3/src/main/java/org/jclouds/s3/blobstore/S3BlobStore.java:193]: This can throw `NullPointerException` when it return `null`.  Can you change the implementation of `getVersions` to return `false` when it is not specified?

File: blobstore/src/main/java/org/jclouds/blobstore/domain/internal/MutableStorageMetadataImpl.java [hunk 3]
```diff
@@ -93,6 +107,16 @@ public void setETag(String eTag) {
...
    }
 
+   @Override
+   public void setVersionId(String versionId) {
+      this.versionId = versionId;
+   }
+
+   @Override
+   public void setIsLatest(String isLatest) {
+   this.isLatest = isLatest;
```
Comment (gaul) [inline blobstore/src/main/java/org/jclouds/blobstore/domain/internal/MutableStorageMetadataImpl.java:117]: indentation

File: blobstore/src/main/java/org/jclouds/blobstore/options/ListContainerOptions.java [hunk 4]
```diff
@@ -231,6 +244,15 @@ public ListContainerOptions delimiter(String delimiterString) {
       return this;
    }
 
+   /**
+    * return a listing of all objects inside the store, recursively.
+    */
+   public ListContainerOptions versions() {
+      // checkArgument(path == null, "path and recursive combination currently not supported");
```
Comment (gaul) [inline blobstore/src/main/java/org/jclouds/blobstore/options/ListContainerOptions.java:251]: commented out code

File: apis/s3/src/main/java/org/jclouds/s3/S3Client.java [hunk 5]
```diff
@@ -640,6 +651,30 @@ void enableBucketLogging(@Bucket @EndpointParam(parser = AssignCorrectHostnameFo
...
+   @Path("/")
+   @ResponseParser(BucketConfigurationHandler.class)
+   String deleteBucketConfiguration(@Bucket @EndpointParam(parser = AssignCorrectHostnameForBucket.class) @BinderParam(
+       BindAsHostPrefixIfConfigured.class) @ParamValidators(BucketNameValidator.class) String bucketName, BucketConfigOptions... options);
+
+   @Named("PutBucketConfiguration")
+   @PUT
+   @Path("/")
+   @ResponseParser(BucketConfigurationHandler.class)
+   String putBucketConfiguration(@Bucket @EndpointParam(parser = AssignCorrectHostnameForBucket.class) @BinderParam(
```
Comment (gaul) [inline apis/s3/src/main/java/org/jclouds/s3/S3Client.java:673]: Can you pull the bucket configuration into a separate PR with unit tests?

File: apis/s3/src/main/java/org/jclouds/s3/config/S3HttpApiModule.java [hunk 6]
```diff
@@ -184,7 +184,7 @@ protected void bindErrorHandlers() {
    }
 
    protected void bindRequestSigner() {
-      bind(RequestAuthorizeSignature.class).to(RequestAuthorizeSignatureV2.class).in(Scopes.SINGLETON);
+      bind(RequestAuthorizeSignature.class).to(RequestAuthorizeSignatureV4.class).in(Scopes.SINGLETON);
```
Comment (gaul) [inline apis/s3/src/main/java/org/jclouds/s3/config/S3HttpApiModule.java:187]: I am willing to entertain changing the default to V4 but this should at least be a separate change.  We also probably need to document how to easily use the older signer.

**Review by gaul — CHANGES_REQUESTED** [review]
Please remove the noisy formatting changes and other commented code.  Also this will be easier to review and merge if you reduce the scope of functionality, e.g., put the bucket configuration in a separate PR.  Finally you need to add some kind of integration tests and probably a few unit tests.

**Review by gaul — CHANGES_REQUESTED** [review]
Still no unit or integration tests.  Please respond to every reviewer comment with "Done" or explain the reason that you have not addressed it.

File: apis/s3/src/main/java/org/jclouds/s3/config/S3HttpApiModule.java (hunk 6 above)
Comment (jiffysrc) [reply apis/s3/src/main/java/org/jclouds/s3/config/S3HttpApiModule.java:187]: Hi @gaul 
We will be raising this change in separate PR as suggested.
and could you please elaborate on how to document for older signer.

File: apis/s3/src/main/java/org/jclouds/s3/S3Client.java (hunk 5 above)
Comment (jiffysrc) [reply apis/s3/src/main/java/org/jclouds/s3/S3Client.java:673]: Hi @gaul 
Will be raising new PR for this change with unit tests, and other noise changes highlighted in above comments.

File: blobstore/src/main/java/org/jclouds/blobstore/options/ListContainerOptions.java (hunk 4 above)
Comment (jiffysrc) [reply blobstore/src/main/java/org/jclouds/blobstore/options/ListContainerOptions.java:251]: Will be removed as part of bucket configuration PR

File: blobstore/src/main/java/org/jclouds/blobstore/domain/internal/MutableStorageMetadataImpl.java (hunk 3 above)
Comment (jiffysrc) [reply blobstore/src/main/java/org/jclouds/blobstore/domain/internal/MutableStorageMetadataImpl.java:117]: Done as part of bucket configuration PR

File: apis/s3/src/main/java/org/jclouds/s3/S3Client.java (hunk 1 above)
Comment (jiffysrc) [reply apis/s3/src/main/java/org/jclouds/s3/S3Client.java:267]: All the noise changes will be corrected with bucket configuration PR

File: apis/s3/src/main/java/org/jclouds/s3/domain/internal/ListVersionsResponseImpl.java [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+        result = prime * result + ((marker == null) ? 0 : marker.hashCode());
+        result = prime * result + maxKeys;
+        result = prime * result + ((name == null) ? 0 : name.hashCode());
+        result = prime * result + ((prefix == null) ? 0 : prefix.hashCode());
+        result = prime * result + (truncated ? 1231 : 1237);
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```
[the remark this answers is not in the document]
Comment (jiffysrc) [reply apis/s3/src/main/java/org/jclouds/s3/domain/internal/ListVersionsResponseImpl.java:153]: Done as part of bucket configuration PR

**Comment by jiffysrc** [comment]
Hi @gaul 
We will be divide the changes in this pr into 2 new separate PR's having,
1 : bucket configuration changes , 
2 : Changing to RequestAuthorizeSignatureV4

and close this PR.

File: apis/s3/src/main/java/org/jclouds/s3/config/S3HttpApiModule.java (hunk 6 above)
Comment (gaul) [reply apis/s3/src/main/java/org/jclouds/s3/config/S3HttpApiModule.java:187]: I think we should do something like this:

https://github.com/gaul/jclouds-1/pull/new/s3/signer

Note that this does not work and needs to move some code around.

@nacx Does this seem like the correct approach?

Comment (jiffysrc) [reply apis/s3/src/main/java/org/jclouds/s3/config/S3HttpApiModule.java:187]: Hi @nacx 
Could you please provide feedback on above approach,

@gaul 
 Will you be completing the remaining change on this PR/move code around.

Comment (nacx) [reply apis/s3/src/main/java/org/jclouds/s3/config/S3HttpApiModule.java]: The approach looks good @gaul. Probably the code moving needs to change to something [like this](https://github.com/apache/jclouds/blob/master/apis/openstack-keystone/src/main/java/org/jclouds/openstack/keystone/auth/config/AuthenticationModule.java#L61-L77). The selected lines are the ones that configure, in OpenStack, the version of the Keystone API to inject based on a property. A similar pattern could be used here.

Comment (gaul) [reply apis/s3/src/main/java/org/jclouds/s3/config/S3HttpApiModule.java]: @nacx Thanks for the hint!

@jiffysrc Please test this -- I only looked at the unit tests.

Comment (jiffysrc) [reply apis/s3/src/main/java/org/jclouds/s3/config/S3HttpApiModule.java]: @gaul 
Tested, This seems fine.

**Comment by jiffysrc** [comment]
Raised Separate PR for Bucket Configuration Options, Hence closing this PR
